scratchpad/tn_api/tn.py

```diff
@@ -0,0 +1,202 @@
+import numpy as np
+import math
+from dataclasses import dataclass
+from typing import TypeVar, Generic, Iterable
+
+class Array(np.ndarray):
+    shape: tuple
+
+D = TypeVar('D') # tensor data type (numpy, torch, etc.)
+
+class ContractionInfo:
+    pass
+
+class TensorNetworkIFC(Generic[D]):
+    def __init__(self, *args, **kwargs):
+        ...
+
+    def optimize(self, out_indices: Iterable = []) -> ContractionInfo:
+        return ContractionInfo()
+    
+    # slice not inplace
+    def slice(self, slice_dict: dict) -> 'TensorNetwork':
+        ...
+
+    # contract to produce a new tensor
+    def contract(self, contraction_info: ContractionInfo) -> D:
+        ...
+
+    # 
+    def copy(self):
+        ...
+
+    def add(self, other: "TensorNetworkIFC[D] | D"):
+        ...
+
+
+    @classmethod
+    def new_random_cpu(cls, dims: Iterable[int])-> 'TensorNetworkIFC[D]':
+        ...
+
+    def __eq__(a, b):
+        ...
+
+
+N = TypeVar('N', bound=np.ndarray)
+
+@dataclass
+class Port:
+    tensor_ref: int
+    ix: int
+
+class TensorNetwork(TensorNetworkIFC[np.ndarray]):
+    tensors: Iterable[np.ndarray]
+    shape: tuple
+    edges: tuple
+
+    def __init__(self, *args, **kwargs):
+        self._tensors = []
+        self._edges = tuple()
+        self.shape = tuple()
+        self._indices = {}
+
+    # slice not inplace
+    def slice(self, slice_dict: dict) -> 'TensorNetwork':
```

Comment:
@danlkv thanks for the previous comment, i believed i've addressed that in this update